File: 백준/Silver/11724. 연결 요소의 개수/연결 요소의 개수.py
N, M = map(int, input().split())

edges = {}

for _ in range(M):
    a, b = map(int, input().split())
    if a in edges: edges[a].append(b)
    else: edges[a] = [b]
    if b in edges: edges[b].append(a)
    else: edges[b] = [a]


visited = [False for _ in range(N)]
component = 0

for node in range(N):
    if not visited[node]:
        stk = [node+1]
        component += 1
        while stk:
            start_node = stk.pop()
            if visited[start_node-1]:
                continue
            visited[start_node-1] = True
            
            if not start_node in edges:
                continue
                
            for edge in edges[start_node]:
                if visited[edge-1]:
                    continue
                stk.append(edge)

print(component)

File: 백준/Silver/11724. 연결 요소의 개수/연결 요소의 개수.cc
#include <iostream>
#include <vector>
#include <queue>

using namespace std;

vector<int> edges[1010];
vector<bool> visited(1010, false);
int cnt = 0;

void DFS(int node) {
    visited[node] = true;

    for (int i : edges[node]) {
        if (!visited[i]) {
            DFS(i);
        }
    }
}

int main() {
    int N, M;
    cin >> N >> M;

    for (int i = 0; i < M; i++) {
        int u, v;
        cin >> u >> v;
        edges[u].push_back(v);
        edges[v].push_back(u);
    }

    for (int i = 1; i <= N; i++) {
        if (!visited[i]) {
            cnt++;
            DFS(i);
        }
    }
    cout << cnt;

    return 0;
}
